src/kv/kvtypes.h

```diff
@@ -116,6 +101,50 @@ namespace kv
     virtual crypto::Sha256Hash get_root() = 0;
   };
 
+  class Replicator
+  {
+  public:
+    struct NodeConf
+    {
+      NodeId node_id;
+      std::string host_name;
+      std::string port;
+    };
+    virtual ~Replicator() {}
+    virtual bool replicate(
+      const std::vector<std::tuple<Version, std::vector<uint8_t>, bool>>&
+        entries) = 0;
+    virtual Term get_term() = 0; // TODO(#api): this ought to have a more
+                                 // abstract name than Term
+
+    virtual Term get_term(Version version) = 0;
+    virtual Version get_commit_idx() = 0;
+
+    virtual NodeId leader() = 0;
+    virtual NodeId id() = 0;
+    virtual bool is_leader() = 0;
+    virtual bool on_request(kv::TxHistory::RequestCallbackArgs args) = 0;
+    virtual void periodic(std::chrono::milliseconds elapsed) = 0;
+
+    virtual bool is_follower() = 0;
+    virtual void recv_message(const uint8_t* data, size_t size) = 0;
+    virtual void add_configuration(
+      Index idx,
+      std::unordered_set<NodeId> conf,
+      const NodeConf& node_conf = {}) = 0;
+
+    virtual void force_become_leader() = 0;
+    virtual void force_become_leader(
+      Version index,
+      Term term,
+      const std::vector<Version>& terms,
+      Version commit_idx_) = 0;
+
+    virtual void enable_all_domains() = 0;
```

Comment:
This is clearly much beyond the scope of a Replicator.

Reply:
moving everything except replicate() into a separate class the Raft and Pbft will inherit from as we speak